File: app/src/main/java/tabletop/velocic/com/worldforgerpgtools/GeneratorDeserializer/TableEntries.java
package tabletop.velocic.com.worldforgerpgtools.GeneratorDeserializer;

import java.util.Map;

public class TableEntries
{
    public String Name;
    public Map<String, String> Metadata;
    public String DiceRange;
    public Map<String, String> RerollSubTable;

    public String getName()
    {
        return Name;
    }
    public Map<String, String> getMetadata()
    {
        return Metadata;
    }
    public String getDiceRange()
    {
        return DiceRange;
    }
    public Map<String, String> getRerollSubTable()
    {
        return RerollSubTable;
    }
    public SubTableRollRange getSubTableRollRange()
    {
        if (RerollSubTable.containsKey("ValidSubTableEntryRange")) {
            String range = RerollSubTable.get("ValidSubTableEntryRange");
            int separatorIndex = range.indexOf("-");
            String minValue = range.substring(0, separatorIndex);
            String maxValue = range.substring(separatorIndex + 1);

            return new SubTableRollRange(
                Integer.parseInt(minValue),
                Integer.parseInt(maxValue)
            );
        }

        return new SubTableRollRange(1, Integer.MAX_VALUE - 1);
    }
    public int getNumSubTableRolls()
    {
        if (RerollSubTable.containsKey("NumSubTableRolls")) {
            int numRerolls = Integer.parseInt(RerollSubTable.get("NumSubTableRolls"));
        }

        return 1;
    }
}

File: app/src/main/java/tabletop/velocic/com/worldforgerpgtools/GeneratorDeserializer/TableEntries.kt
package tabletop.velocic.com.worldforgerpgtools.GeneratorDeserializer

import com.google.gson.annotations.SerializedName

class TableEntries(
    @SerializedName("Name")
    val name: String,

    @SerializedName("Metadata")
    val metadata: Map<String, String>,

    @SerializedName("DiceRange")
    private val diceRangeString: String,

    @SerializedName("RerollSubTable")
    val rerollSubTable: Map<String, String>?
) {
    val diceRange
        get() = parseDiceRangeString(diceRangeString)

    fun getSubTableRollRange() : IntRange {
        if (rerollSubTable == null) {
            return 1 until Int.MAX_VALUE
        }

        if (rerollSubTable.containsKey("ValidSubTableEntryRange")) {
            return parseDiceRangeString(rerollSubTable["ValidSubTableEntryRange"])
        }

        return 1 until Int.MAX_VALUE
    }

    fun getNumSubTableRolls() : Int
    {
        if (rerollSubTable == null) {
            return 0
        }

        return rerollSubTable["NumSubTableRolls"]?.toInt() ?: 1
    }

    private fun parseDiceRangeString(diceRangeString: String?) : IntRange {
        val separatorIndex = diceRangeString?.indexOf("-") ?: -1

        val (left, right) = if (separatorIndex == -1) {
            val singleItemRange = diceRangeString?.toInt() ?: 1
            Pair(singleItemRange, singleItemRange)
        } else {
            Pair(
                diceRangeString?.substring(0, separatorIndex)?.toInt() ?: 1,
                diceRangeString?.substring(separatorIndex + 1)?.toInt() ?: Int.MAX_VALUE - 1
            )
        }

        val (min, max) = if (left < right) {
            Pair(left, right)
        } else {
            Pair(right, left)
        }

        return min..max
    }
}